Feature: Setup local database

CR1337 opened the pull request:
Fixes #61

We now have local databases for dev and test and a script to manage it. Also we have an abstract BaseModel all Models can inherit from.

See [Wiki](https://github.com/BP2022-AP1/bp2022-ap1/wiki/Database-Management)

## PR checklist

- [x] Acceptance criteria fulfilled
- [x] Additional features are tested
- [x] Docs (code, wiki & diagrams) updated
- [ ] Breaking changes anounced
- [x] Dev-branch has been merged into local branch to resolve conflicts
- [x] Tests and linter have passed AFTER local merge
- [ ] Another dev reviewed and approved

CR1337 committed:
installed peewee and marshmallow

CR1337 committed:
installed postgres driver

CR1337 committed:
feat: implemented BaseModel

CR1337 committed:
feat: Script for managing local database

CR1337 committed:
feat: script for testing

CR1337 committed:
feat: setup local databases in docker

CR1337 committed:
feat: implemented basic database test

CR1337 committed:
style: reformatted

CR1337 committed:
fix: fixed linting errors

CR1337 committed:
fix: fixed linting errors

CR1337 committed:
fix: fixed linting error

CR1337 committed:
docs: added explanatory comment

CR1337 committed:
fix: fixed linting error

CR1337 committed:
fix: linting error

CR1337 commented:
The test pipeline fails because there is no database defined yet. This should be done next.

CR1337 commented:
> The test pipeline fails because there is no database defined yet. This should be done next.

Or should this be done inside this PR so that we don't break the pipeline an dev?

CR1337 committed:
style: reformatted

antonykamp commented on src/base_model.py:
```diff
@@ -0,0 +1,29 @@
+from peewee import Model, PostgresqlDatabase
+
+
+db: PostgresqlDatabase = PostgresqlDatabase(database="postgres",user="postgres",password="root",host="localhost")
```
We should use environments variables here, because in the future we will have a production database with a better password 👍 An `.env` file is quite common.

antonykamp commented on db/local/db.py:
```diff
@@ -0,0 +1,167 @@
+        print(f'Rolled back migration: {name}')
+
+
+def execute_on_args():
```
Please add some documentation with available commands and arguments to this file 😇 This would help using it in the future

antonykamp commented on db/local/db.py:
```diff
@@ -0,0 +1,167 @@
+def run(args: argparse.Namespace):
+    """Run the database
+
+    :param args: parsed arguments
```
Instead of `parsed arguments` as description, could you please add the content of args (in this case `database_name`). Otherwise this comment wouldn't help much 😇 same with stop, recreate, migration.

antonykamp commented on src/base_model.py:
```diff
@@ -0,0 +1,29 @@
+db: PostgresqlDatabase = PostgresqlDatabase(database="postgres",user="postgres",password="root",host="localhost")
+
+
+class BaseModel(Model):
```
I really love the idea of the `BaseModel`! It's a good way to combine marshmellow and peewee quickly! 💚

antonykamp commented on db/local/db.py:
```diff
@@ -0,0 +1,167 @@
+def recreate(args: argparse.Namespace):
+    """Drop and create database tables
+
+    :param args: parsed arguments
```
Could you please add a hint, that `recreate` doesn't need arguments, but has to forward it to `drop` and `create`

antonykamp commented on src/base_model.py:
```diff
@@ -0,0 +1,34 @@
+import marshmallow as marsh
+from peewee import Model, PostgresqlDatabase
+
+db: PostgresqlDatabase = PostgresqlDatabase(
```
We should use environments variables here, because in the future we will have a production database with a better password 👍 An .env file is quite common.

antonykamp commented on tests/db/test_db.py:
```diff
@@ -0,0 +1,17 @@
+    test_value = IntegerField()
+
+
+def test_db():
```
Please add a more descriptive name 😇 e.g. `test_db_connection_workflow`

antonykamp commented on db/local/db.py:
```diff
@@ -0,0 +1,173 @@
+models: list[Type[BaseModel]] = BaseModel.__subclasses__()
+
+
+def run(args: argparse.Namespace):
```
Question: Do we need to validate `args` or does argparse take over the job?

antonykamp replied:
Maybe we should flag some arguments as required

antonykamp commented on bin/test:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
```
We could think replacing the bash scripts with 
https://github.com/nat-n/poethepoet

antonykamp reviewed (changes requested):
Hey 👋 
I really love the structure and the idea with the `BaseModel` class 💚 Good job!
I just requested some styling and the use of environment variables.
Additionally I proposed some styling changes, nothing to worry about 😄

antonykamp commented:
> > The test pipeline fails because there is no database defined yet. This should be done next.
>
> Or should this be done inside this PR so that we don't break the pipeline an dev?

I think it's a good idea to add an empty database in this PR 👍

antonykamp committed:
Merge branch 'dev' into feature/61-setup-local-database

antonykamp committed:
doc: specify parameter `args` more

Specify parameter `args` in run, stop, migration and recreate more.

antonykamp committed:
style: rename test_db test

Rename test_db test to test_db_connection_workflow

antonykamp committed:
fix: make arguments mandatory

Make arguments of the command line tool mandatory.

antonykamp committed:
Revert "fix: make arguments mandatory"

This reverts commit 4ba1015da7355350b330897fd6d9a4b489b85b47.

antonykamp committed:
feat: replace bashfiles with poe commands

Replacing the bash files with poe commands enables the developer to work with the DB over an established interface

antonykamp committed:
feat: replace DB configuration with env-variables

Replaces the database configuration values with environment variables.
It's more secure.

antonykamp commented:
review and merge after https://github.com/BP2022-AP1/bp2022-ap1/pull/74

SaturnHafen commented on pyproject.toml:
```diff
@@ -22,6 +22,11 @@ pytest = "^7.2.2"
 black = "^23.1.0"
 isort = "^5.12.0"
 pylint = "^2.17.0"
+traci = "^1.16.0"
+peewee = "^3.16.0"
+peewee-migrate = "^1.6.6"
+marshmallow = "^3.19.0"
```
Why are these dev-dependencies?

antonykamp replied:
Good question 👍 We should discuss what a dev dependency defines. A dev dependency doesn't directly impact the product, e.g., linting or command definition. I would add this definition to the wiki if it's fine.

SaturnHafen commented on db/local/db.py:
```diff
@@ -0,0 +1,173 @@
+    print("Stopping database...")
+    subprocess.run(
+        [
+            "docker-compose",
```
Depending on your docker install, this could also be `docker compose`. `docker-compose` is the older version (See: <https://docs.docker.com/compose/compose-v2/#differences-between-compose-v1-and-compose-v2>)

antonykamp replied:
Good catch! Should we use v2?

SaturnHafen replied:
Well v1 will be deprecated end  of june 2023, so yes.

antonykamp replied:
👍

SaturnHafen commented on db/local/db.py:
```diff
@@ -0,0 +1,173 @@
+    print(os.getcwd())
+    subprocess.run(
+        [
+            "docker-compose",
```
Same as below

SaturnHafen commented on db/local/db.py:
```diff
@@ -0,0 +1,173 @@
+    print("Stopping database...")
+    subprocess.run(
+        [
+            "docker-compose",
```
This also assumes that you are member of the `docker` group (linux), as otherwise you will not have access to docker without `sudo`

antonykamp replied:
Good catch too! We should add this to the documentation, that you should be part of the docker group.

SaturnHafen commented on db/local/db.py:
```diff
@@ -0,0 +1,173 @@
+
+MIGRATION_DIRECTORY: str = "db/local/migrations"
+
+models: list[Type[BaseModel]] = BaseModel.__subclasses__()
```
Is that how it is supposed to work? Seems kind of brittle...

antonykamp replied:
It works 😄 But we can think about declaring an array by hand. This would require changing the array, when declaring a new subclass of `BaseModel`.

SaturnHafen replied:
I guess we can leave it as-is.

SaturnHafen reviewed (commented):
Looks good so far!
We need to document the dependencies on `docker` (and `poetry`, but that might be for a later pr)

antonykamp committed:
Merge branch 'dev' into feature/61-setup-local-database

antonykamp committed:
fix: replace "docker-compose" with "docker compose"

"docker-compose" will be deprecated soon. Replace with "docker compose".

antonykamp committed:
fix: move dependencies to main

Move peewee, peewee-migrate, marshmallow, psycopg2 to main dependencies.

antonykamp commented:
approved? @SaturnHafen

SaturnHafen reviewed (approved):
Looks good to me!
